rfcs/2020-11-06-4909-remap-template-fields.md

```diff
@@ -0,0 +1,142 @@
+# RFC 4909 - 2020-11-06 - Remap support for template strings
+
+We would like to unify the templating syntax within configuration fields with
+the Remap language.
+
+
+## Scope
+
+This RFC will look into ways we can use Remap whilst still supporting the current
+method of templating fields.
+
+
+## Motivation
+
+Using the Remap language rather than simple template fields provides much greater
+power in how these fields are defined. A difficulty does arise in that the 
+templatie fields did provide a list of fields that were used in the template. 
+This was straightforward as the template fields were very simply defined. With
+Remap it is much harder to come up with this list.
+
+The Loki sink wants a list of the fields that were used in the template. It
+uses this list to remove these fields from the event sent to Loki. 
+
+As an example if you had this sink:
+
+```toml
+[sinks.loki]
+  type = "loki"
+  inputs = ["json"]
+  endpoint = "http://localhost:3100"
+  remove_label_fields = true
+  labels.key = "{{foo}}-{{buzz}}"
+```
+
+And this event was sent to it:
+
+```json
+{"foo": "bar", "buzz": "zab", "message1": "thing", "message2": "thong"}
+```
+
+The actual message sent to Loki would be:
+
+```
+{"message1": "thing", "message2": "thong"}
+```
+
+With the label `key = bar-zab` attached to the message. The fields `foo` and 
+`buzz` that were used in defining the label have been removed from the message.
+
+With Remap this list becomes dynamic. So the script `if .foo { .bar } else 
+{ .baz }` means sometimes `.bar` is used and other times `.baz` is used. It is 
+not clear how we should proceed to determine which fields need to be removed
+from the message.
+
+
+## Internal Proposal
+
+As there are a number of complications with working out which fields are used
+in the output, we could only support this feature if the old style template
+syntax is used. If Remap is used, the user won't have the option to remove these
+fields from the message in the Loki sink.
```

Comment:
I actually have on my list a change to the `del` function for it to return the value of the last field it deletes (this is part of an effort to get rid of the `None` return value in Remap, which brings with it a whole bunch of tricky to catch issues).

Given that, it would be possible for someone to write:

```
labels.key = "{{ del(.foo) }}-{{ del(.buz) }}"
```

To achieve the same result, using Remap.

Reply:
Even without that change, it would still be possible (buy way less ergonomic) to achieve the same result with the current Remap feature-set:

```lua
labels.key = """{{
$foo = .foo
del(.foo)

$buz = .buz
del(.buz)

$foo + "-" + $buz
}}
"""
```